### DISCUSSION TITLE
Is it possible to make greenlet optional dependency?

### DISCUSSION THREAD

**Discussion by heckad** [sqlalchemy/sqlalchemy#7092] [Ideas]
In some ouwer prohjects we doesn't use async code and greenlet. It would be grate if greenlet became optional. Is it possible?

**Comment by zzzeek** [comment]
hi -

greenlet only installs by default on platforms for which there is a python wheel file available on pypi and should not have any negative side effects.  Please see #6136.     if you want to install SQLAlchemy with no dependencies you can use `--no-deps` with pip. 

there's unfortunately no standard way in pip that im familiar with to exclude specific dependencies for packages - in https://stackoverflow.com/questions/33441033/pip-install-to-custom-target-directory-and-exclude-specific-dependencies there are various recipes to create custom requirements.txt files and install from that.  if you have very specific installation requirements such that the presence of the greenlet package is taking up too much space or something (such as on mobile or embedded) I would recommend using specific requirements.txt files.

**Reply by heckad** [reply]
What if create package `sqlalchemy[async]` wich will install sqlalchemy with greenlet? This can be achieved by add optional dependency.

**Reply by CaselIT** [reply]
there is already that extra requires. I think the idea was to avoid requiring additional setup to enable async usage

**Reply by zzzeek** [reply]
we do have that package name set up.  but we tried to make the default what we felt the vast majority of users would want.  if we didnt have it in default we'd get constant complaints about it.   

still missing here is why is having greenlet installed a problem.  the largest size for .whl I can see on pypi is 150K, which is quite tiny compared to SQLAlchemy coming in at 1.5M.

**Reply by heckad** [reply]
> we do have that package name set up. but we tried to make the default what we felt the vast majority of users would want. if we didnt have it in default we'd get constant complaints about it.

You can do to try import greenlet and if import error then print error with command wich users can use for install sqlalchemy with greenlets. Now we use pipenv and they can't protect from install `greenlet`.  

> still missing here is why is having greenlet installed a problem. the largest size for .whl I can see on pypi is 150K, which is quite tiny compared to SQLAlchemy coming in at 1.5M.

Unfortunately, on our platform greenlet fails during build. Now, to update sqlalchemy to version 1.4, we need to switch to requirements.txt in all projects, since sqlalchemy is a dependency in the integration library.

**Reply by CaselIT** [reply]
> Unfortunately, on our platform greenlet fails during build. Now, to update sqlalchemy to version 1.4, we need to switch to requirements.txt in all projects, since sqlalchemy is a dependency in the integration library.

what is the `platform_machine` that is reported in your platform?

having dependencies sure is a pain...

**Reply by heckad** [reply]
> what is the `platform_machine` that is reported in your platform?

docker image python:3.9.5-alpine. Today was fixed. 

Is it possible to extend the condition so that in pipenv you can specify that you don't need to install greenlet?

**Reply by CaselIT** [reply]
>  python:3.9.5-alpine

it does not seem possible to detect alpine using the platform, see https://github.com/sqlalchemy/sqlalchemy/issues/6136#issuecomment-810190341

Personally I would advise against alpine with python, as I mentioned here https://github.com/sqlalchemy/sqlalchemy/issues/6136#issuecomment-808454873, since it usually cannot install compiled wheels

> Is it possible to extend the condition so that in pipenv you can specify that you don't need to install greenlet?

if you can find a way I think we can change it sure

**Reply by bigcat88** [reply]
How can I install SqlAlchemy without greenlet? They have no binary wheels for musl aarch64, and user have no root rights.
If I provide `--no-deps` what deps must be installed to make SQLAlchemy work?(except pymysql and pg8000)

**Reply by bigcat88** [reply]
Never-mind, will PR to greenlet to add build wheel for musl aarch64.

**Reply by CaselIT** [reply]
> what deps must be installed to make SQLAlchemy work?

from here https://github.com/sqlalchemy/sqlalchemy/blob/c98c9d7133d6a3eb1b92a95a565d704bcbbcaa4e/setup.cfg#L42-L44 the only other requirement in 1.4 is `importlib-metadata` in python < 3.8
